packages/beacon-node/src/network/network.ts

```diff
@@ -243,22 +244,53 @@ export class Network implements INetwork {
       const blocks = await this.reqResp.beaconBlocksByRange(peerId, request);
       const blobsSidecars = await this.reqResp.blobsSidecarsByRange(peerId, request);
 
-      if (blocks.length !== blobsSidecars.length) {
-        throw Error(`blocks.length ${blocks.length} != blobsSidecars.length ${blobsSidecars.length}`);
-      }
-
       const blockInputs: BlockInput[] = [];
+      let blobSideCarIndex = 0;
+      let lastMatchedSlot = -1;
+
+      // Match blobSideCar with the block as some blocks would have no blobs and hence
+      // would be omitted from the response. If there are any inconsitencies in the
+      // response, the validations during import will reject the block and hence this
+      // entire segment.
+      //
+      // Assuming that the blocks and blobs will come in same sorted order
       for (let i = 0; i < blocks.length; i++) {
         const block = blocks[i];
-        const blobsSidecar = blobsSidecars[i];
-
-        // TODO EIP-4844: Do more verification blob is for block
-        if (block.message.slot !== blobsSidecar.beaconBlockSlot) {
-          throw Error(`blob does not match block slot ${block.message.slot} != ${blobsSidecar.beaconBlockSlot}`);
+        let blobsSidecar: eip4844.BlobsSidecar;
+
+        if (blobsSidecars[blobSideCarIndex]?.beaconBlockSlot === block.message.slot) {
+          blobsSidecar = blobsSidecars[blobSideCarIndex];
+          lastMatchedSlot = block.message.slot;
+          blobSideCarIndex++;
+        } else {
+          // Quick inspect if the blobsSidecar was expected
+          const blobKzgCommitmentsLen = (block.message.body as eip4844.BeaconBlockBody).blobKzgCommitments.length;
+          if (blobKzgCommitmentsLen !== 0) {
+            throw Error(
+              `Missing blobsSidecar for blockSlot=${block.message.slot} with blobKzgCommitmentsLen=${blobKzgCommitmentsLen}`
+            );
+          }
+          blobsSidecar = {
+            beaconBlockRoot: this.config.getForkTypes(block.message.slot).BeaconBlock.hashTreeRoot(block.message),
+            beaconBlockSlot: block.message.slot,
+            blobs: [],
+            kzgAggregatedProof: ckzg.computeAggregateKzgProof([]),
```

Comment:
Why do you need to create an empty BlobsSidecar?

Reply:
because empty BlobsSidecar may be skipped in the req/resp, thats whats happening in devnet3